config/application.rb

```diff
@@ -25,6 +25,17 @@ class Application < Rails::Application
       config.paths["app/views"].unshift ::Configuration.custom_views_root.to_s
     end
 
+    # Handle case where app database file is missing and user does not use the Dashboard to launch app
+    if (! ::Configuration.database_path.file?) || ::Configuration.database_path.empty?
+      p 'Building the database'
+      require 'rake'
+      FileUtils.chdir ::Configuration.dataroot.parent do
+        load_tasks
+        ::Configuration.production_database_path.parent.mkpath
+        Rake::Task['db:setup'].invoke
```

Comment:
In this block, why not execute `bin/setup-production` instead in a bundle clean block i.e. https://github.com/OSC/ood-dashboard/blob/9fe5a286e9582d2f14bc7ec2d3fcdf6ffc16fc8e/app/apps/ood_app.rb#L197-L223? And the question about handling failure as well... perhaps this should be the first thing in config.ru that is executed and based on the result (a failure or not) we either load this app or load an app without it.

Even better would be to decouple the connection to the database from the app initialization. I believe this is possible, though it would take some time understanding the details of how ActiveRecord works.

Reply:
Oh, its because the `bin/setup-production` script is what currently fails in this instance? Perhaps that means that `bin/setup-production` has a defect as well? If we have this block execute `bin/setup-production` and then fix that script so when launching from the dashboard it works as well, maybe that is a better solution.

Reply:
As written `bin/setup-production` assumes that if the database file exists then it is valid. If the file does not exist then it runs `RAILS_ENV=production rake db:setup` as a subprocess which is inadequate.

Ideally we want to handle:
- File exists and is a valid SQLite3 database
- File does not exist at all
- File exists but is invalid:
    - File is zero length
    - File is reported as invalid according to SQLite3's [`PRAGMA integrity_check`](https://www.sqlite.org/pragma.html#pragma_integrity_check)

Where file could be either the development database (I delete mine all the time (but I usually remember how to recover from this error)) or the production one that users keep deleting.